datalad/distribution/install.py

```diff
@@ -43,6 +43,131 @@ def _with_sep(path):
     return path + sep if not path.endswith(sep) else path
 
 
+def _install_subds_from_flexible_source(ds, sm_path, sm_url, recursive):
+    """Tries to obtain a given subdataset from several meaningful locations"""
+    # shortcut
+    vcs = ds.repo
+    repo = vcs.repo
+    # compose a list of candidate clone URLs
+    clone_urls = []
+    # if we have a remote, let's check the location of that remote
+    # for the presence of the desired submodule
+    tracking_branch = repo.active_branch.tracking_branch()
+    remote_url = ''
+    # remember suffix
+    url_suffix = ''
+    if tracking_branch:
+        # name of the default remote for the active branch
+        remote_name = repo.active_branch.tracking_branch().remote_name
+        remote_url = vcs.git_get_remote_url(remote_name, push=False)
+        if remote_url.endswith('/.git'):
+            url_suffix = '/.git'
+            remote_url = remote_url[:-5]
+        # attempt: submodule checkout at parent remote URL
+        clone_urls.append('{0}/{1}{2}'.format(
+            remote_url, sm_path, url_suffix))
+    # attempt: configured submodule URL
+    if sm_url.startswith('/') \
+            or sm_url.split('://')[0] in ('http', 'https', 'ssh', 'file', 'git'):  # XXX: should we allow ANY protocol?
+        # this seems to be an absolute location -> take as is
+        clone_urls.append(sm_url)
+    else:
+        # need to resolve relative URL
+        if not remote_url:
+            # we have no remote URL, hence we need to go with the
+            # local path
+            remote_url = ds.path
+        remote_url_l = remote_url.split('/')
+        sm_url_l = sm_url.split('/')
+        for i, c in enumerate(sm_url_l):
+            if c == '..':
+                remote_url_l = remote_url_l[:-1]
+            else:
+                clone_urls.append('{0}/{1}{2}'.format(
+                    '/'.join(remote_url_l),
+                    '/'.join(sm_url_l[i:]),
+                    url_suffix))
+                break
+    # now loop over all candidates and try to clone
+    subds = Dataset(opj(ds.path, sm_path))
+    for clone_url in clone_urls:
+        lgr.debug("Attempt clone of subdataset from: {0}".format(clone_url))
+        try:
+            subds = Install.__call__(
+                dataset=subds, path=None, source=clone_url,
+                recursive=recursive, add_data_to_git=False)
+        except GitCommandError:
```

Comment:
Not sure. Wouldn't we have to check if there is anything there before we start cloning? If there is something, what should we do? Fail before starting?

Also we probable need an `annex_remove` implementation to do this right ;-)

Let's postpone this change. I will add a note for now.